cpp/src/arrow/acero/source_node.cc

```diff
@@ -406,7 +407,7 @@ struct SchemaSourceNode : public SourceNode {
 struct RecordBatchReaderSourceNode : public SourceNode {
   RecordBatchReaderSourceNode(ExecPlan* plan, std::shared_ptr<Schema> schema,
                               arrow::AsyncGenerator<std::optional<ExecBatch>> generator)
-      : SourceNode(plan, schema, generator) {}
+      : SourceNode(plan, schema, generator, Ordering::Implicit()) {}
```

Comment:
Isn't this a bit too restrictive? This means all RecordBatch reader source **always** preserve the implicit order, even if if users do not care? Shouldn't this be configurable? What about other sources? Shouldn't the behaviour be the same for any source?

Reply:
Probably option to specify ordering would be best. Or maybe automatically assign implicit ordering based on 
`batch.index!=compute::kUnsequencedIndex`?

Reply:
Can you elaborate on why your asof_join enhancement requires specifically the `RecordBatchReaderSrouceNode` to have implicit ordering?

Reply:
It does not. I assumed that when user provides a generator,  it has some kind of implicot order.  Now I see it is not necessarily a case.